src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/SimpleTaskQueue.cs

```diff
@@ -22,41 +22,25 @@ internal sealed class SimpleTaskQueue
         private readonly object _gate = new object();
 
         private Task _latestTask;
-        private int _taskCount;
 
         public SimpleTaskQueue(TaskScheduler taskScheduler)
         {
             _taskScheduler = taskScheduler;
-
-            _taskCount = 0;
             _latestTask = Task.CompletedTask;
         }
 
-        private TTask ScheduleTaskWorker<TArg, TTask>(Func<int, TArg, TTask> taskCreator, TArg arg)
-            where TTask : Task
-        {
-            lock (_gate)
-            {
-                _taskCount++;
-                var delay = (_taskCount % 100) == 0 ? 1 : 0;
-
-                var task = taskCreator(delay, arg);
-
-                _latestTask = task;
-
-                return task;
-            }
-        }
-
         [PerformanceSensitive(
             "https://developercommunity.visualstudio.com/content/problem/854696/changing-target-framework-takes-10-minutes-with-10.html",
             AllowCaptures = false)]
         [SuppressMessage("Style", "VSTHRD200:Use \"Async\" suffix for async methods", Justification = "This is a Task wrapper, not an asynchronous method.")]
         public Task ScheduleTask(Action taskAction, CancellationToken cancellationToken = default)
         {
-            return ScheduleTaskWorker(
-                (delay, arg) => arg.Item1._latestTask.ContinueWithAfterDelay(arg.taskAction, arg.cancellationToken, delay, TaskContinuationOptions.None, arg.Item1._taskScheduler),
-                (this, taskAction, cancellationToken));
+            lock (_gate)
+            {
+                var task = _latestTask.SafeContinueWith(_ => taskAction(), cancellationToken, TaskContinuationOptions.None, _taskScheduler);
```

Comment:
Why the intermediate? Just wrote _latestTask = _latestTask and then return later? You've got a lock acquired either way...

Reply:
sure

Reply:
Actually, that only works with the first overload. Other overloads return `Task<T>`, not `Task`